### PR TITLE
Convert Guardcondition to use C++ classes

### PR DESCRIPTION
**Pull request by ahcorde**
Related to https://github.com/ros2/rclpy/issues/665

This converts the GuardCondition functions to a C++ class. 

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

File: rclpy/src/rclpy/guard_condition.hpp [hunk 1]
```diff
@@ -17,31 +17,73 @@
...
+  {
+    return rcl_guard_condition_.get();
+  }
+
+  /// Handle destructor for guard condition
+  static void
+  _rclpy_destroy_guard_condition(void * p)
+  {
+    (void)p;
+    // Empty destructor, the class should take care of the lifecycle.
```
Comment (hidmic) [inline rclpy/src/rclpy/guard_condition.hpp:62]: @ahcorde what ensures this object won't get trashed before the capsule does?

File: rclpy/src/rclpy/guard_condition.hpp [hunk 2]
```diff
@@ -17,31 +17,73 @@
...
+
+  /// Get rcl_client_t pointer
+  rcl_guard_condition_t * rcl_ptr() const
+  {
+    return rcl_guard_condition_.get();
+  }
+
+  /// Handle destructor for guard condition
+  static void
+  _rclpy_destroy_guard_condition(void * p)
```
Comment (hidmic) [inline rclpy/src/rclpy/guard_condition.hpp:59]: @ahcorde nit^2: make this `private`

**Review by hidmic — APPROVED** [review]

File: rclpy/src/rclpy/guard_condition.hpp (hunk 1 above)
Comment (ahcorde) [reply rclpy/src/rclpy/guard_condition.hpp:62]: The Python class `SignalHandlerGuardCondition(GuardCondition)` is the only one in the code making use of this reference. In the destructor I can see how the capsule is destroyed first and then the object.

https://github.com/ros2/rclpy/blob/7dff0ee8bc6f299b46a6d0c2f8adc5cd518ae95f/rclpy/rclpy/signals.py#L38-L42

or should we include some kind of logic ?

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14390/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9164/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12065/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14488/)

File: rclpy/src/rclpy/guard_condition.hpp (hunk 1 above)
Comment (hidmic) [reply rclpy/src/rclpy/guard_condition.hpp]: I see. This is arguably a (temporary?) implementation detail. No need to change it.

**Review by hidmic — APPROVED** [review]

**Pull request merged**